java/ql/src/experimental/Security/CWE/CWE-555/CredentialsInPropertiesFile.ql

```diff
@@ -0,0 +1,87 @@
+/**
+ * @name Cleartext Credentials in Properties File
+ * @description Finds cleartext credentials in Java properties files.
+ * @kind problem
+ * @id java/credentials-in-properties
+ * @tags security
+ *       external/cwe/cwe-555
+ *       external/cwe/cwe-256
+ *       external/cwe/cwe-260
+ */
+
+import java
+import semmle.code.configfiles.ConfigFiles
+
+private string suspicious() {
+  result = "%password%" or
+  result = "%passwd%" or
+  result = "%account%" or
+  result = "%accnt%" or
+  result = "%credential%" or
+  result = "%token%" or
+  result = "%secret%" or
+  result = "%access%key%"
+}
+
+private string nonSuspicious() {
+  result = "%hashed%" or
+  result = "%encrypted%" or
+  result = "%crypt%"
+}
+
+/** Holds if the value is not cleartext credentials. */
+bindingset[value]
+predicate isNotCleartextCredentials(string value) {
+  value = "" // Empty string
+  or
+  value.length() < 7 // Typical credentials are no less than 6 characters
+  or
+  value.matches("% %") // Sentences containing spaces
+  or
+  value.regexpMatch(".*[^a-zA-Z\\d]{3,}.*") // Contain repeated non-alphanumeric characters such as a fake password pass**** or ????
+  or
+  value.matches("@%") // Starts with the "@" sign
+  or
+  value.regexpMatch("\\$\\{.*\\}") // Variable placeholder ${credentials}
+  or
+  value.matches("%=") // A basic check of encrypted credentials ending with padding characters
+  or
+  value.matches("ENC(%)") // Encrypted value
+  or
+  value.toLowerCase().matches(suspicious()) // Could be message properties or fake passwords
```

Comment:
What are message properties? Perhaps just "looks like a filler value"?

Reply:
Message properties look like <code>messages.properties</code> under the <code>test</code> directory. An example is:
```
login.password_expired=Your current password has expired. Please reset your password.
```
The key contains the keyword "password" while the value is a message to be displayed on the UI instead of credentials therefore shall be excluded. 

Java applications mainly use properties files for two purposes - configuration properties and localized text messages to be displayed on the UI. For example, an application can have a file <code>messages.properties</code> for English messages while have another file <code>messages_fr.properties</code> for the same messages in French. 

The query tries to identify message properties and excludes them from returned results to reduce FPs.

Reply:
Makes sense. Suggest adding a slightly longer comment then giving examples of what you intend to exclude

Reply:
Added a longer comment